[doc] C-API documentation & generator (#5736)
taichi-dev/taichi: docs/lang/articles/c-api/taichi_core.md

## Availability

Taichi C-API intends to support the following backends:

|Backend|Offload Target|Maintenance Tier|
|-|-|-|
|Vulkan|GPU|Tier 1|
|CUDA (LLVM)|GPU (NVIDIA)|Tier 1|
|CPU (LLVM)|CPU|Tier 1|
|OpenGL|GPU|Tier 2|
|DirectX 11|GPU (Windows)|N/A|
|Metal|GPU (macOS, iOS)|N/A|

The backends with tier-1 support are being developed and tested more intensively. And most new features will be available on Vulkan first because it has the most outstanding cross-platform compatibility among all the tier-1 backends.
For the backends with tier-2 support, you should expect a delay in the fixes to minor issues.

For convenience, in the following text and other C-API documents, the term *host* refers to the user of the C-API; the term *device* refers to the logical (conceptual) compute device, to which Taichi's runtime offloads its compute tasks. A *device* may not be a physical discrete processor other than the CPU and the *host* may *not* be able to access the memory allocated on the *device*.

Unless otherwise specified, **device**, **backend**, **offload target**, and **GPU** are interchangeable; **host**, **user code**, **user procedure**, and **CPU** are interchangeable.

> ## How to...

Reviewer: ```suggestion
## HowTo
```

mai.host_write = TI_TRUE;
mai.usage = TI_MEMORY_USAGE_STORAGE_BIT;
TiMemory steaming_memory = ti_allocate_memory(runtime, &mai);

// ...

std::vector<uint8_t> src = some_random_data_source();

void* dst = ti_map_memory(runtime, steaming_memory);
std::memcpy(dst, src.data(), src.size());
ti_unmap_memory(runtime, streaming_memory);
```

To read data back to the host, `host_read` *must* be set to [`TI_TRUE`](#definition-ti_true).

```cpp
TiMemoryAllocateInfo mai {};
mai.size = 1024; // Size in bytes.
mai.host_read = TI_TRUE;
mai.usage = TI_MEMORY_USAGE_STORAGE_BIT;
TiMemory read_back_memory = ti_allocate_memory(runtime, &mai);

// ...

std::vector<uint8_t> dst(1024);
void* src = ti_map_memory(runtime, read_back_memory);
std::memcpy(dst.data(), src, dst.size());
ti_unmap_memory(runtime, read_back_memory);

ti_free_memory(runtime, read_back_memory);
```

> You can set `host_read` and `host_write` at the same time.

### Load and destroy a Taichi AOT module

You can load a Taichi AOT module from the filesystem.

```cpp
TiAotModule aot_module = ti_load_aot_module(runtime, "/path/to/aot/module");
```

`/path/to/aot/module` should point to the directory that contains a `metadata.tcb`.

> You can destroy an unused AOT module if you have done with it; but please ensure there is no kernel or compute graph related to it pending to [`ti_submit`](#function-ti_submit).

Reviewer: ```suggestion
You can destroy an unused AOT module, but please ensure that there is no kernel or compute graph related to it pending to [`ti_submit`](#function-ti_submit).
```
- Am not quite certain about "If you have done with it". Grammatically it ought to be "if you are done with it". I believe it is a given, so I omitted it. 
- Do not use conj. together with semicolons.
